### ISSUE TITLE
MeshCentral Assistant: files download doesn't work edge/chrome

### ISSUE THREAD

**Issue by energywave** [Ylianst/MeshCentral#5659]
**Describe the bug**
When connected to a MeshCentral assistant Windows computer file download doesn't work. When you connect to the file section and click on a file link to download it nothing happens.
It's not a client device issue because if I connect to whatever agent is connected I can correctly download files using the same procedure.

**To Reproduce**
1. Run an assistant on a Windows PC
2. Connect to the file section
3. Click on a file link to download it (nothing happens...)

**Expected behavior**
Click on a file link -> download prompt from the browser and ability to locally save the file.

**Server Software (please complete the following information):**
 - OS: Windows 2012
 - Virtualization: Hyper-V
 - Network: LAN/WAN (no differences)
 - Version: 1.1.16
 - Node: 16.15.1

**Client Device (please complete the following information):**
 - Device: Desktop PC
 - OS: Windows 10
 - Network: Local / WAN (no differences)
 - Browser: Chrome 120.0.6099.130 64bit / Edge 120.0.2210.91 64bit

**Remote Device (please complete the following information):**
 - Device: Desktop PC
 - OS: Windows 10 / Windows 11 (same behavior)
 - Network: Local / WAN (no differences)
 - Current Core Version (if known): v0.1.8175.21599

**Comment by si458** [comment]
works ok here but im using firefox,
have just tried edge and that doesnt work?
weird?

**Comment by si458** [comment]
pushed a temp fix for you!

the issue is because the first byte that gets sent by the agents SHOULD be a command string, which tells the server to set headers for files names etc,

but for some reason the meshassistant is sending file data as the first byte instead of the command string, so then when the next byte comes in and its the command string, it cant write the headers to tell the browser the filename etc because its already send part of the file data!

so the temp fix simply tells the browser the link will download a file and this is its file name
(this was already set for firefox but not chrome/edge due to a bug but it seems the bug might be fixed now)

**Comment by energywave** [comment]
> pushed a temp fix for you!

Hey, I don't have words to thanks your effort... Tomorrow I'll try to change that line and I'll tell you if it will work correctly. In the meanwhile... all my gratitude to have gone deep investigating. I'm happy that Meshcentral is still alive with people that understands it's code!

**Comment by si458** [comment]
No worries! Don't forget to sponsor! https://github.com/si458 (all links down left hand side!)

**Comment by energywave** [comment]
Thank you again for your time! I've tested it and, sorry, it don't works. I've made the same changes in the default.handlebars that you made (The element.setAttribute line is at line 18554 as I suspect the source on the repo has some improvements not yet released...), restarted meshcentral, made an hard refresh and tested again to download a file when connected to an assistant. Same behavior, nothing happens.
Using DevTools I can see the devicefile.ashx coming with the contents in the reply section but i don't see the filename anywhere in the reply headers.
Maybe I have to do something else? I'm missing something to do?

**Comment by si458** [comment]
Are u using minify? 
Is so disable that, as the minify isn't done on the fly

**Comment by energywave** [comment]
I've thinked the same when I've seen the `default-min.handlebars` file so, yes, I was using minify. Now I've set it to false in the config.json file, restarted the service, done an hard refresh again in edge but... no improvements.
What I'm doing wrong...?

**Comment by si458** [comment]
> I've thinked the same when I've seen the `default-min.handlebars` file so, yes, I was using minify. Now I've set it to false in the config.json file, restarted the service, done an hard refresh again in edge but... no improvements. What I'm doing wrong...?

Erm... nothing? So long as u have change the latest commit like for like, it should work? As it works here?
Will have to look when I get stationary.

It's annoying because the bug is actually the meshassistant exe itself! And I can't fix that, so I'm trying to figure out a temp fix for the client side instead

**Comment by energywave** [comment]
> So long as u have change the latest commit like for like, it should work? As it works here? 

Sorry, didn't understood that :(
My meshcentral is at last released version 1.1.16 (I use the update inside the web UI of meshcentral). Is there a way to check that the code I've changed is propagated correctly? I'm not used to node.js and I didn't get if that js is running on server side or on the client side.

> It's annoying because the bug is actually the meshassistant exe itself! And I can't fix that, so I'm trying to figure out a temp fix for the client side instead

The meshassistant exe is written in .net right? Where is the source of that? I could not find it. Maybe, if you tell me what I have to change, I can try to help...

**Comment by si458** [comment]
the standard update methods wont work, so the line you should of changed should be around line 18954
the is one way to update/try if you comfy with it
1. stop meshcentral,
2. BACKUP YOUR WHOLE FOLDER/SERVER!
3. `cd /yourmeshfolder`
4. `npm install Ylianst/MeshCentral` <- this will install the MASTER BRANCH, so it will include extra fixes for other stuff
5. wait about 5mins for it to reinstall from github repo
6. start meshcentral
7. check again
8. **if you want to roll back do the same but instead of `Ylianst/MeshCentral` put `meshcentral@1.1.6`**

**Comment by si458** [comment]
are you still having this issue? can u try the latest 1.1.19 version at all?

**Comment by si458** [comment]
are you still having this issue? plz try the latest 1.1.21 update

**Comment by energywave** [comment]
Sorry that I didn't test it before, my work went crazy... I just tested with ver 1.1.21 and I could donwload ad upload correctly, thank you! :)
I noticed that there are still some glitches when you try to upload an existing file or a file that is in use. The operation terminates silently and there is no error message that let te user understand it. That's true even with the meshagent.
If you want I can open a new issue about that.
